infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java

```diff
@@ -152,14 +164,20 @@ public SessionManager(@Nonnull TopologyDescriptor topology, CorfuRuntime corfuRu
             .build();
         this.localCorfuEndpoint = lrNodeLocator.toEndpointUrl();
         this.metadataManager = new LogReplicationMetadataManager(corfuRuntime, topology.getTopologyConfigId());
-        this.configManager = new LogReplicationConfigManager(runtime);
-        this.upgradeManager = null;
+        this.configManager = new LogReplicationConfigManager(runtime, topology.getLocalClusterDescriptor().getClusterId());
+        this.clientConfigListener = new LogReplicationClientConfigListener(this, configManager, corfuStore);
         this.replicationContext = new LogReplicationContext(configManager, topology.getTopologyConfigId(), localCorfuEndpoint);
         this.replicationManager = replicationManager;
         this.router = router;
         this.incomingMsgHandler = logReplicationServer;
     }
 
+    /**
+     * Start client config listener from discovery service upon leadership acquired.
+     */
+    public void startClientConfigListener() {
```

Comment:
Just a thought: Wondering sessionManager is the right place for clientConfigListener? Did you explore adding it in config manager?

Reply:
From what I saw, SessionManager is a more "central" or "heavy" object and ConfigManager is initialized here. The clientConfigListener needs to access both and it's kind of a mediator, like in the workflow of group-destination change, it needs to go to configManger to get impacted sessions, and come back to SessionManager to enqueue FORCED_SNAPSHOT_SYNC events. And also the listener is supposed to start at very beginning, so I thought the proper place to initialize it is here

Reply:
Also on roleChange to SOURCE right? Just wondering how that workflow operates ?

Reply:
It somehow came from the dependency of those classes: SessionManager is the heavier one and it's the only one among those created directly in discoveryService. We create ConfigManager inside SessionManager so it serves as something like the registry of all the configs from all the sessions. ConfigListener needs dependency on both and it's something like a coordinator/mediator of those two - it receives client config updates and update config through ConfigManager, received registration info and indicate SessionManager to create new sessions, etc.